### ISSUE TITLE
[MNG-6323] Deadlock in multithreaded dependency resolution

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7217]
**[Ben Caradoc-Davies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bencaradocdavies)** opened **[MNG-6323](https://issues.apache.org/jira/browse/MNG-6323?redirect=false)** and commented

Maven 3.5.2 multithreaded builds experience a deadlock not seen with Maven 3.5.0.

To reproduce the issue, clone GeoServer:

```
git clone https://github.com/geoserver/geoserver.git
cd geoserver
```

Build GeoServer community modules with:

```
mvn -f src/community/pom.xml -B -T4 -U -Prelease -PcommunityRelease -DskipTests clean install
```

Builds that normally take 2-4 minutes instead experience long hangs. `jstack` output (attached) suggests a deadlock (two different hangs attached). Some of the locks are in `TIME_WAIT` and eventually the build completes after 30-45 minutes, but this is enough to cause builds on Travis to be killed for their failure to output for ten minutes. (Travis upgraded to Maven 3.5.2 a week ago.)

I have only seen the failures with -U. The hang does not occur in single-threaded builds. There are no "*.lock" files in the local repository during the hang so the deadlocks are not mediated by the filesystem. CPU utilisation is zero suggesting a deadlock not a livelock.

See also discussion on the geoserver-devel mailing list: http://osgeo-org.1560.x6.nabble.com/Travis-failures-started-with-new-trusty-images-td5346836.html

---

**Affects:** 3.5.2

**Attachments:**
- [geoserver-community-maven-hang-jstack.txt](https://issues.apache.org/jira/secure/attachment/12902694/geoserver-community-maven-hang-jstack.txt) (_19.23 kB_)
- [geoserver-community-maven-hang-jstack-2.txt](https://issues.apache.org/jira/secure/attachment/12902693/geoserver-community-maven-hang-jstack-2.txt) (_28.10 kB_)

**Issue Links:**
- [MNG-6418](https://issues.apache.org/jira/browse/MNG-6418) Parallel builds with empty local repository hang
 (_**"is duplicated by"**_)
- [MNG-6324](https://issues.apache.org/jira/browse/MNG-6324) multithread build: DefaultProjectDependenciesResolver.resolve waits for a lock indenfinetly causing the build to hang
 (_**"is duplicated by"**_)
- [MNG-6361](https://issues.apache.org/jira/browse/MNG-6361) Multithreaded Build hangs for 10+ Minutes, then resumes
 (_**"is duplicated by"**_)
- [MRESOLVER-39](https://issues.apache.org/jira/browse/MRESOLVER-39) deadlock during multithreaded dependency resolution

1 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Ben Caradoc-Davies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bencaradocdavies)** commented

By the way, reverting MNG-6025 (be223808939d75152d7157db33d45f230114555e) did not fix. I did not try reverting the upgrade to Resolver 1.1.0 (MNG-6258).

**Comment by jira-importer** [comment]
**[Ben Caradoc-Davies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bencaradocdavies)** commented

Another Travis user reports that they were also affected by this bug and able to work around it by downgrading to Maven 3.5.0: https://github.com/travis-ci/travis-ci/issues/8932

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

MNG-6324 seems to have a reproducable project. Then it is a matter of git-bisect to find the trouble making commit.

**Comment by jira-importer** [comment]
**[Zoltan Haindrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kgyrtkirk)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) If you haven't started bisecting this yet...I can set it up ; and probably give you a commit hash which have caused this problem...

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I would really appreciate it if you could help finding the first bad commit for us.

**Comment by jira-importer** [comment]
**[Zoltan Haindrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kgyrtkirk)** commented

This was a very tricky thing...I think there might be probably another issue as well...because these builds always started with empty repo, I think that might trigger also proabilistically trigger another problem: after some point the build breaks down because it's unable to locate some snapshot artifact - usually it lists one artifact from the actually built project;  this is probably less critical; so for now I've just masked this problem away....

The commit which causes this issue is:

```
commit e520324c8d45e2465cf8aeafb4d57cef30250136
    [MNG-6258] Upgrade to Maven Resolver 1.1.0
```

[Ben Caradoc-Davies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bencaradocdavies) congrats on guessing the right commit :)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

So one of the following changes seems to have more impact than expected: https://issues.apache.org/jira/projects/MRESOLVER/versions/12339212

This looks like another bisect challenge, but now you need to replace the Maven Artifact Resolver SNAPSHOT in the existing Maven distribution.

**Comment by jira-importer** [comment]
**[Zoltan Haindrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kgyrtkirk)** commented

probably...I'll continue localizing in that project...these tickets don't seem to be related to multi threading
I'm already realigning this thing to track it down in the resolver :)

**Comment by jira-importer** [comment]
**[Zoltan Haindrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kgyrtkirk)** commented

narrowed it down to ad50215d27feede0ad0e5eb83ae96c6d6fdcc639
which have changed the files appearing in the stacktraces;
I've opened MRESOLVER-39 to let the maven-resolver project members know about this issue

**Comment by jira-importer** [comment]
**[Henrik Lundahl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henrik.lundahl)** commented

MRESOLVER-39 has been fixed now, but Maven Resolver 1.1.1 has not yet been released.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [91d1edf14e0ed198c917efeddd0b8241980ec0ed](http://git-wip-us.apache.org/repos/asf/maven/commit/91d1edf1)
